### ISSUE TITLE
Add CONTRIBUTING.md

### ISSUE THREAD

**Issue by AliSoftware** [SwiftGen/SwiftGen#135]
- Mention the Moya Contributors Guidelines and the fact that I wish to give push access to every contributor
- Document a bit the organization of the project: `swiftgen-cli` for the parsing of arguments (using `Commander`) and calling `GenumKit`, `GenumKit` for all the logic parsing the files and creating some Stencil contexts from that, `Stencil` for template rendering, …
- Document the steps to contribute: e.g. for every new template, don't forget to update Unit Tests too, for every new feature don't forget to include a Playground page, …
- Improve code documentation to help contributors understand the codebase better.

**Comment by AliSoftware** [comment]
Perfect timing: We'll likely base that `CONTRIBUTING.md` file and all the related documentation and CoC on https://github.com/Moya/contributors/pull/7 :+1:

**Comment by Dschee** [comment]
I'm **all in for a CONTRIBUTING.md** file since I just happened to have problems or at least questions when I started to work on #93.

I think the file should also include **a section on how to setup the project** from the contributor side. Here's **what I have done** and **what issues I faced** and am still facing:

- make sure Homebrew and CocoaPods are installed and up-to-date
- run `brew install swiftlint` (I had it installed, but this is what should be in the docs)
- run `pod install` in the root folder

**My first issue** was that `pod install` lead to the following warnings:

```
[!] The `swiftgen [Debug]` target overrides the `EMBEDDED_CONTENT_CONTAINS_SWIFT` build setting defined in `Pods/Target Support Files/Pods-swiftgen/Pods-swiftgen.debug.xcconfig'. This can lead to problems with the CocoaPods installation
    - Use the `$(inherited)` flag, or
    - Remove the build settings from the target.

[!] The `swiftgen [Release]` target overrides the `EMBEDDED_CONTENT_CONTAINS_SWIFT` build setting defined in `Pods/Target Support Files/Pods-swiftgen/Pods-swiftgen.release.xcconfig'. This can lead to problems with the CocoaPods installation
    - Use the `$(inherited)` flag, or
    - Remove the build settings from the target.
```

My CocoaPods version is 1.0.1. These warnings confused me – I can't tell if they are of any relevance.

**My second issue** was after opening `SwiftGen.xcworkspace` and hitting `Cmd+B` to build the swiftgen-cli target. Xcode built for some time but then **failed** with the following output:

```
PhaseScriptExecution ⚠️\ SwiftLint /Users/Dschee/Library/Developer/Xcode/DerivedData/SwiftGen-hefihlopydloiferyjpxynfkjfmz/Build/Intermediates/SwiftGen.build/Debug/swiftgen.build/Script-09E690541D037CB80024BE5F.sh
    cd "/Users/Dschee/Code/GitHub/OS X/SwiftGen"
    /bin/sh -c /Users/Dschee/Library/Developer/Xcode/DerivedData/SwiftGen-hefihlopydloiferyjpxynfkjfmz/Build/Intermediates/SwiftGen.build/Debug/swiftgen.build/Script-09E690541D037CB80024BE5F.sh

Unrecognized arguments: "X/SwiftGen/swiftgen-cli"
Unrecognized arguments: "X/SwiftGen/GenumKit"
Command /bin/sh failed with exit code 1
```

**My third issue** appeared right after thinking "maybe I'm not supposed to build this myself, some script may take care of it" and went for `Cmd+U` to run the tests. This lead to a long list of warnings and errors within the Issue Navigator.

**The warnings** came from swiftlint stating the following for different files:

```
X/SwiftGen/UnitTests/expected/Colors-File-CustomName.swift.out:1: warning: Trailing Newline Violation: Files should have a single trailing newline. (trailing_newline)
```

**Most of the errors** came from a lovely animal stating the following for different files:

```
cat: X/SwiftGen/UnitTests/expected/Colors-Empty.swift.out: No such file or directory
```

One additional error stated the following:

```
Command /bin/sh emitted errors but did not return a nonzero exit code to indicate failure
```

The funny part about all this is that Xcode showed me both the **"Build Succeeded" and the "Tests Succeeded"** info views so that I'm now completely confused. ^.^

Do you know what I missed and clarify this for me and others in the future via the CONTRIBUTING.md file? That would be ✨ 👍 

Thanks in advance!

**Comment by AliSoftware** [comment]
Thanks 💯 for the feedback @Dschee ! Definitely useful to know what issues you encountered so we can document them better and help future contributors, thanks a lot 👌

**Comment by AliSoftware** [comment]
To answers your questions:

SwiftGen is a bit special indeed, because it is a command line tool (not a `.app`) but is written in Swift. 
What this means is:

* It doesn't have an app bundle wrapping the binary executable, frameworks and resources like regular ~~OS X~~ macOS `.app` bundles, but instead builds a standalone binary executable
* But being written in Swift, it needs to be linked against the Swift dynamic libraries (various `libswiftXXX.dylib` libraries) and also has other dependencies (like `Stencil` & `Commander` frameworks) and needs to be linked against them too

So about your issues:

### `EMBEDDED_CONTENT_CONTAINS_SWIFT` warning

The `EMBEDDED_CONTENT_CONTAINS_SWIFT` build setting in Xcode is used to tell Xcode to copy those various `libswitXXX.dylib` dynamic libraries into the final `.app` application Bundle. But since we don't have such a bundle in our special case, Xcode would fail trying to do that. That's why I had to disable this build setting.

CocoaPods warns about it (and is right to do so) because in normal setups you're not supposed to do that. But in my special case I need to disable it, and perform the linking myself because those libraries will be next to the binary (installed by Homebrew alongside the binary) but not in a `.app`

So disabling that `EMBEDDED_CONTENT_CONTAINS_SWIFT` is intentional, and the warning is normal because we do something special, and thus it can be ignored in our very special case.

### Trying to build the swift-cli target

As you probably guessed, this one is tied to the exact same reason. As of Xcode 7 (and even Xcode 8), Xcode doesn't support targets building **Command-Line binaries** that **link to external libraries** (like the Swift dylibs) all by itself. that's why I had to handle that by hand.

Sadly, that means that I use Xcode merely only as an IDE to type my Swift code but can't actually use Xcode to **build** the final command-line binary since it's not an usual `.app` but still links to external libs. Instead, you need to use the `Rakefile` to build the command line binary (as you may have guessed from the comments in your 3rd issue)… using `rake install` in the terminal, and all you can do with Xcode is type the code and run the Test target. That's also why the test target is so important

### Warnings

This was an issue with my scripts that tried to validate the generated `*.swift.out` files — output expected by the Unit Tests as the generated code — so that we ensure that the generated code (which, from the POV of SwiftGen and Stencil is actually just text generated from a template) is valid Swift and compiles.
So that we don't end up shipping SwiftGen templates that actually generate invalid code or code that doesn't compile, and also (and that's the point of [the SwiftLint tool](https://github.com/realm/SwiftLint/blob/master/README.md)) to validate the generated code also follows the coding guidelines and is consistent.

My script intended to `cat` each `*.swift.out` file and pipe that content to `swiftlint` so it can validate that the generated code follows the expected coding guidelines. But obviously the `cat` command seemed to fail on your machine — probably because you have a space ` ` somewhere in your path leading to the SwiftGen project and I forgot to quote that path in my script — hence the warnings you saw.

I _think_ I fixed that recently so this problem in my scripts should have been solved by now, though.


HTH

**Comment by Dschee** [comment]
Thank you @AliSoftware for the quick and detailed response. 👍 

Okay, just to pipe your answers into something useful and because it is the topic of this issue anyway I've tried to put my understandings of all of this into a CONTRIBUTING.md file section and created a merge request: #147.

Also note that I renamed my folder "OS X" to "macOS" (which I needed to do sometime anyway) and now all warnings and errors when building disappeared (except for two "Todo Violations" from swiftlint which are okay). So 👍 for the whitespace pointer!
